api/types/database.go

```diff
@@ -637,6 +655,55 @@ func (d *DatabaseV3) CheckAndSetDefaults() error {
 	return nil
 }
 
+// checkAndSetDynamoDBDefaults checks DynamoDB configuration and sets defaults.
+func (d *DatabaseV3) checkAndSetDynamoDBDefaults() error {
+	if d.Spec.AWS.AccountID == "" {
+		return trace.BadParameter("database %q AWS account ID is empty", d.GetName())
+	}
+
+	region, err := awsutils.DynamoDBRegionForEndpoint(d.Spec.URI)
+	switch {
+	case err != nil:
+		// when region parsing returns an error but the region is set, it's ok because we can just construct the URI using the region,
+		// so we check if the region is configured to see if this is really a configuration error.
+		if d.Spec.AWS.Region == "" {
+			// the AWS region is empty and we can't derive it from the URI, so this is a config error.
+			return trace.BadParameter("database %q AWS region is empty and cannot be derived from the URI %q",
+				d.GetName(), d.Spec.URI)
+		}
+		if awsutils.IsAWSEndpoint(d.Spec.URI) {
+			// The user configured an AWS URI that which doesn't look like a DynamoDB endpoint.
+			// The URI must look like <service>.<region>.<partition> or <region>.<partition>
+			return trace.Wrap(err)
+		}
```

Comment:
I have been conflicted about this as well. One thing I've realized is that fips endpoints will not work currently, since they will get rewritten to the standard endpoint like `dynamodb-fips.us-west-2.amazonaws.com` -> `dynamodb.us-west-2.amazonaws.com`. At the same time, we determine the endpoint based on the incoming request at run time... I may need to rework the configuration logic such that the following behavior holds:
1. blank uri is derived from the region and set to the standard endpoint suffix.
2. custom uri is allowed and we just try to derive region from the uri if we need to and can do so.
3. don't override any custom uri. But in the engine, check to see if the configured uri is an aws endpoint missing the service subdomain and add the subdomain if that is the case.

In this configuration logic, if the user sets the uri like:
`URI: "dynamodb.us-west-2.amazonaws.com"` then the service subdomain will not be overwritten and any requests sent to that uri which are trying to do streams operations or DAX will just fail. This is something we can document in the guide, as in "don't specify the URI if you want to allow dax/streams/dynamodb requests to be proxied by the agent. If you want fips, then you need to specify the fips uri".

Reply:
The alternative to this would be to force users to setup a static database for each service endpoint and for each region they want. I think that's not very friendly UX, but I'm open to doing that if others think that is better

Reply:
> 1. blank uri is derived from the region and set to the standard endpoint suffix.
> 2. custom uri is allowed and we just try to derive region from the uri if we need to and can do so.
> 3. don't override any custom uri. But in the engine, check to see if the configured uri is an aws endpoint missing the service subdomain and add the subdomain if that is the case.

I think this is reasonable.  `region` with blank URI is likely the "default" way ppl would use most of the time. And it's fine to ask users to specify a custom URI to overwrite it when the "default" way doesn't work. 

The only part I find a little confusing personally is having the `.us-west-2.amazonaws.com` as the URI (you would also see this in `tsh db ls -v` too). One other alternative is to use a custom schema when there is a blank URI (for example, populate the URI to something like `aws://dynamodb`. Krzysztof did something similar for the Azure CLI support for app access. This way, you can tell the difference between a blank URI vs a custom URI, and if need to, force the user to specify a full/valid URI when not doing the "default" way.

> The alternative to this would be to force users to setup a static database for each service endpoint and for each region they want.

I think this is also fine since this is the case for other databases today. Personally, I do like the above-quoted logic better.

Reply:
Ok I've updated the uri logic as described in my initial suggestion above.

Reply:
Cool. Do we still need this `if awsutils.IsAWSEndpoint(d.Spec.URI)` check?